record-and-playback/presentation/scripts/publish/presentation.rb

```diff
@@ -580,251 +545,202 @@ def events_parse_shape(shapes, event, current_presentation, current_slide, times
       return
     end
-    if $version_atleast_2_0_0
+    if @version_atleast_2_0_0
       shape[:thickness_percent] = thickness.text.to_f
     else
       shape[:thickness] = thickness.text.to_i
     end
   end
-  if  shape[:type] == 'rectangle' or
-      shape[:type] == 'ellipse' or shape[:type] == 'triangle'
-    fill = event.at_xpath('fill')
-    fill = fill.nil? ? "false" : fill.text
+  if %w[ellipse rectangle triangle].include?(shape_type)
+    fill = event.at_xpath('fill')&.text || 'false'
     shape[:fill] = fill =~ /true/ ? true : false
   end
-  if shape[:type] == 'rectangle'
-    square = event.at_xpath('square')
-    if !square.nil?
-      shape[:square] = (square.text == 'true')
-    end
-  end
-  if shape[:type] == 'ellipse'
-    circle = event.at_xpath('circle')
-    if !circle.nil?
-      shape[:circle] = (circle.text == 'true')
-    end
-  end
-  if shape[:type] == 'pencil'
-    commands = event.at_xpath('commands')
-    if !commands.nil?
-      shape[:commands] = commands.text.split(',').map { |c| c.to_i }
-    end
-  end
-  if shape[:type] == 'poll_result'
+
+  case shape_type
+  when 'rectangle'
+    square = event.at_xpath('square')&.text
+    shape[:square] = (square == 'true') if square
+  when 'ellipse'
+    circle = event.at_xpath('circle')&.text
+    shape[:circle] = (circle == 'true') if circle
+  when 'pencil'
+    commands = event.at_xpath('commands')&.text
+    shape[:commands] = commands.split(',').map(&:to_i) if commands
+  when 'poll_result'
     shape[:num_responders] = event.at_xpath('num_responders').text.to_i
     shape[:num_respondents] = event.at_xpath('num_respondents').text.to_i
     shape[:result] = event.at_xpath('result').text
-  end
-  if shape[:type] == 'text'
+  when 'text'
     shape[:text_box_width] = event.at_xpath('textBoxWidth').text.to_f
     shape[:text_box_height] = event.at_xpath('textBoxHeight').text.to_f
 
-    calced_font_size = event.at_xpath('calcedFontSize')
-    unless calced_font_size
-      BigBlueButton.logger.warn("Draw #{shape[:shape_id]} Shape #{shape[:shape_unique_id]} ID #{shape[:id]} is missing calcedFontSize")
+    unless (calced_font_size = event.at_xpath('calcedFontSize'))
+      BigBlueButton.logger.warn("Draw #{draw_id} Shape #{shape[:shape_unique_id]} ID #{shape_id} is missing calcedFontSize")
       return
     end
 
     shape[:calced_font_size] = calced_font_size.text.to_f
 
     shape[:font_color] = color_to_hex(event.at_xpath('fontColor').text)
-    text = event.at_xpath('text')
-    if !text.nil?
-      shape[:text] = text.text
-    else
-      shape[:text] = ''
-    end
+    shape[:text] = event.at_xpath('text')&.text || ''
   end
 
   # Find the previous shape, for updates
   prev_shape = nil
-  if !shape[:id].nil?
+  if shape_id
     # If we have a shape ID, look up the previous shape by ID
-    prev_shape = shapes.find_all {|s| s[:id] == shape[:id] }.last
+    # Don't look for updates if the drawing has ended
```

Comment:
Thank you, this change has been undone, keeping the search starting from the last element.
When animations are turned off, so to speak, is the `prev_shape` still necessary? Since every new addition to the whiteboard then corresponds to the `DRAW_END` case. The performance gains of this PR mostly came from skipping this search so it would be unfortunate to see it go

Reply:
The problem is that we need some way to detect whether or not shape updates happen in the recording. There's a couple of options I can see for optimizing this going forwards:

1. We can update the recording version number, then use version matching to skip the shape update searching. (Basically, assume BBB ≥ 2.5 will not do animated shape events in recordings).
2. We can optimize the shape lookup by maintaining a list of "open" shapes (which would hold pointers to shapes which have received START/UPDATE events but no END events) so it doesn't have to do a search through the full list of shapes.